Initial client/server networking

caspernorrbin opened the pull request:
Adds client/server networking within `admirals::net`

Uses ASIO as a submodule, so it needs it needs to be downloaded like the rest of the submodules.

The communication is message-based, with messages from `Message.hpp` being sent. The whole implementation uses templates to allow for user-defined message types. This allows messages to be of any type and be of any size.


Each connection from a client or the server is handled by a connection object from `Connection.hpp`. Each network connection has two connection objects associated with it: One on the client, and one on the server. The functionality is broadly the same for both, with some small differences in specific functions.

Message queues are used for messaging. Each Connection object has an outgoing message queue, and all clients and the server all have queues for incoming messages. The clients/server can thus choose when to read these queues to process incoming messages.

Since the implementation builds on templates, the implementation cannot be in source files. To circumvent this, the implementation is placed in `.tpp` files (source files for templates), with them being included at the bottom of the header. I am very open to potential other solutions for this.


Also included are two test files: One server and one client. This implements a small "game" of an 8x8 grid. Clients can press 'Q' in their console to spawn a "ship" at (0,0), press 'W' to move it to a random location, and press 'E' to quit. The game waits for 2 players to connect before automatically starting. If run on the same computer, both clients will read the same input and both will try to send an action to the server.


resolves #9, resolves #11

JoXW100 reviewed (commented):
I feel like the lambda expressions in `Connection.tpp` are difficult to read, and we should extract them to functions.

I also feel like the additional complexity of using enums for `Message::id` is not worth it, by removing the use of enums here, most of the network implementation can be done without using generic classes. 

I have created a branch [networking_without_enum](https://github.com/joelsiks/admirals/tree/networking_without_enum)  where the enum is removed and `.tpp` files are changed into `.cpp` files.

joelsiks commented on src/engine/net/Message.hpp:
```diff
@@ -0,0 +1,60 @@
+    template <typename DataType>
+    friend Message<T> &operator<<(Message<T> &msg, const DataType &data) {
+        // Check that the data type is trivially copyable
+        static_assert(std::is_standard_layout<DataType>::value,
+                      "Data is too complex to be pushed into vector");
```
Smart!

joelsiks commented on src/engine/net/Server.hpp:
```diff
@@ -0,0 +1,69 @@
+        return false;
+    }
+    // Called when a client disconnects
+    virtual void OnClientDisconnect(std::shared_ptr<Connection<T>> client) {}
+    // Called when a message is received
+    virtual void OnMessage(std::shared_ptr<Connection<T>> client,
+                           Message<T> &message) {}
```
These should probably be purely virtual, i.e `virtual void OnClientDisconnect(std::shared_ptr<Connection<T>> client) = 0;`

caspernorrbin replied:
Fixed in 13a5840

joelsiks reviewed (commented):
I also think that the lambda functions are somewhat hard to read. One alternative is to move the lambda function into its own function right above the associated function it is used in. That way the deep indentation isn't a problem any more.

caspernorrbin commented on src/engine/net/Client.tpp:
```diff
@@ -0,0 +1,60 @@
+    try {
+        asio::ip::tcp::resolver resolver(m_io_context);
+        asio::ip::tcp::resolver::results_type endpoints =
+            resolver.resolve(host, port);
```
`resolve()` returns a list of 0 or more endpoints, and can potentially have more than one endpoint. When later connecting, one endpoint is used and referenced.

caspernorrbin commented on src/engine/net/Client.hpp:
```diff
@@ -0,0 +1,44 @@
+    asio::io_context m_io_context;
+    std::thread m_context_thread;
+    std::unique_ptr<Connection<T>> m_connection;
+    MessageQueue<OwnedMessage<T>> m_incoming_messages;
```
Fixed in a839b8f

caspernorrbin commented on src/engine/net/Connection.hpp:
```diff
@@ -0,0 +1,83 @@
+    void WriteValidation();
+    void ReadValidation(admirals::net::Server<T> *server = nullptr);
+
+private:
```
It is a remnant of a previous implementation, removed in 13a5840

caspernorrbin commented on src/engine/net/Connection.hpp:
```diff
@@ -0,0 +1,83 @@
+
+    // Used when validating a connection
+    Server<T> *m_server = nullptr;
+    void (Server<T>::*m_validation_function)(std::shared_ptr<Connection<T>>) =
```
Changed to a `std::function` in 13a5840

caspernorrbin commented on src/engine/net/Connection.tpp:
```diff
@@ -0,0 +1,256 @@
+    void (Server<T>::*validation_function)(std::shared_ptr<Connection<T>>))
+    : m_io_context(io_context), m_socket(std::move(socket)),
+      m_incoming_messages(incoming_messages), m_owner_type(parent),
+      m_server(server), m_validation_function(validation_function) {
```
Moved most to the initializations to the constructor body instead in 13a5840

caspernorrbin commented:
Rebased the [networking_without_enum](https://github.com/joelsiks/admirals/tree/networking_without_enum) branch onto this one to remove templates as done by @JoXW100 .

The read/write lambda functions have been separated and moved directly above the respective functions. I am unsure of where to place these, and am open to suggestions. Additionally, the smaller fixes have been added to further increase readability.

JoXW100 commented on src/engine/net/Connection.cpp:
```diff
@@ -0,0 +1,266 @@
+    asio::async_read(
+        m_socket,
+        asio::buffer(m_messageBuffer.body.data(), m_messageBuffer.body.size()),
+        [this](std::error_code ec, std::size_t length) { HandleReadBody(ec); });
```
Here aswell...

JoXW100 commented on src/engine/net/Connection.cpp:
```diff
@@ -0,0 +1,266 @@
+    asio::async_write(
+        m_socket,
+        asio::buffer(&m_outgoingMessages.Front().header, sizeof(MessageHeader)),
+        [this](std::error_code ec, std::size_t length) {
+            HandleWriteHeader(ec);
+        });
```
Here aswell...

JoXW100 commented on src/engine/net/Connection.cpp:
```diff
@@ -0,0 +1,266 @@
+    asio::async_write(m_socket,
+                      asio::buffer(m_outgoingMessages.Front().body.data(),
+                                   m_outgoingMessages.Front().body.size()),
+                      [this](std::error_code ec, std::size_t length) {
+                          HandleWriteBody(ec);
+                      });
```
Here aswell...

JoXW100 commented on src/engine/net/Connection.cpp:
```diff
@@ -0,0 +1,266 @@
+void Connection::WriteValidation() {
+    asio::async_write(m_socket,
+                      asio::buffer(&m_validationOut, sizeof(uint64_t)),
+                      [this](std::error_code ec, std::size_t length) {
+                          HandleWriteValidation(ec);
+                      });
```
Here aswell...

JoXW100 commented on src/engine/net/Connection.cpp:
```diff
@@ -0,0 +1,266 @@
+// the connection is a server)
+void Connection::ReadValidation() {
+    asio::async_read(m_socket, asio::buffer(&m_validationIn, sizeof(uint64_t)),
+                     [this](std::error_code ec, std::size_t length) {
+                         HandleReadValidation(ec);
+                     });
```
Here aswell...

JoXW100 commented on src/engine/net/Connection.hpp:
```diff
@@ -0,0 +1,87 @@
+
+    // Used when validating a connection
+    Server *m_server = nullptr;
+    // void (Server::*m_validationFunction)(std::shared_ptr<Connection>) =
+    // nullptr;
```
Remove this comment

caspernorrbin replied:
Fixed in 713165b

JoXW100 commented on src/engine/net/Server.cpp:
```diff
@@ -0,0 +1,121 @@
+            std::make_shared<Connection>(
+                Connection::Owner::SERVER, m_ioContext, std::move(socket),
+                m_incomingMessages, this,
+                [this](std::shared_ptr<Connection> connection) {
+                    this->OnClientValidated(connection);
+                });
```
Here aswell...

JoXW100 commented on src/engine/net/Server.cpp:
```diff
@@ -0,0 +1,121 @@
+
+void Server::WaitForClientConnection() {
+    m_acceptor.async_accept(
+        [this](std::error_code ec, asio::ip::tcp::socket socket) {
+            HandleAcceptedConnection(ec, std::move(socket));
+        });
```
Here aswell...

JoXW100 commented on src/engine/net/Connection.cpp:
```diff
@@ -0,0 +1,266 @@
+    const asio::ip::tcp::resolver::results_type &endpoints) {
+    asio::async_connect(
+        m_socket, endpoints,
+        [this](std::error_code ec, asio::ip::tcp::endpoint endpoint) {
+            if (!ec) {
+                // Read the validation number from the server and send it back
+                // hashed
+                ReadValidation();
+            } else {
+                std::cout << "Connect failed: " << ec.message() << std::endl;
+            }
+        });
```
I think we should extract this lambda expression. This one is not that badly formatted, but i think we should stay away from lambda expressions with multiline bodies.

caspernorrbin replied:
Changed in 713165b

JoXW100 reviewed (changes requested):
Looks good, but we should remove most of the leftover lambda expressions. Did you figure out the cause of the warnings?

I have implemented this in branch `networking_bound_handlers` that can be merged into this if you agree with these changes. The changes can be viewed in Draft #40.
